### PR TITLE
tracing: Add basic VSOCK tracing

### PR DESCRIPTION
**Pull request by jodh-intel**
Implement an openTelemetry custom exporter that sends trace spans to a VSOCK socket. A VSOCK-to-span converter (such as the Kata trace forwarder) needs to be running on the host to allow systems like Jaeger to capture the trace spans.

By default, tracing is not enabled (meaning a NOP tracer is used). To activate tracing, set the `agent.kata.enable_tracing=true` in the configuration file.

The type of tracing this change introduces is "static isolated" tracing. See [1] for further details.

> **Note:**
>
> This change only provides the foundational changes for agent tracing work. The feature is _not_ yet complete since it does
> not yet show the correct trace hierarchy.

Fixes: #60.

[1] - https://github.com/kata-containers/agent/blob/master/TRACING.md

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

### CODE REVIEW

**Comment by jodh-intel** [comment]
@fidencio - as discussed yesterday in the AC meeting, here's the basic agent tracing PR. Since I've ripped out most of the trace points, hopefully this branch won't get conflicted too quickly too. :smile:

/cc @cmaf, @egernst.

File: src/agent/src/main.rs [hunk 1]
```diff
@@ -199,6 +204,17 @@ async fn real_main() -> std::result::Result<(), Box<dyn std::error::Error>> {
         ttrpc_log_guard = Ok(slog_stdlog::init().map_err(|e| e)?);
     }
 
+    if config.tracing != tracer::TraceType::Disabled {
+        let _ = tracer::setup_tracing(NAME, &logger, &config)?;
```
Comment (cmaf) [inline src/agent/src/main.rs:208]: I noticed `instrument` is used with functions called prior to setting up tracing. Is this intentional?

File: src/agent/src/tracer.rs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+    let config = Config::default();
+
+    let builder = opentelemetry::sdk::trace::TracerProvider::builder()
+        .with_simple_exporter(exporter)
+        .with_config(config);
+
+    let provider = builder.build();
+
+    let version = None;
```
Comment (cmaf) [inline src/agent/src/tracer.rs:73]: Should `version` have a value?

**Review by cmaf — COMMENTED** [review]
@jodh-intel I'm not a rust expert so I won't explicitly approve this PR, but I left a couple questions.

File: src/trace-forwarder/src/handler.rs [hunk 3]
```diff
@@ -78,7 +81,9 @@ fn handle_trace_data(
 
         reader
             .read_exact(&mut encoded_payload)
-            .map_err(|e| format!("failed to read payload: {:}", e))?;
+            .with_context(|| format!("failed to read payload"))?;
+
+        //.map_err(|e| Err(anyhow!("failed to read payload: {:?}", e)))?;
```
Comment (liubin) [inline src/trace-forwarder/src/handler.rs:86]: Nit: this comment should be deleted?

File: src/agent/src/main.rs (hunk 1 above)
Comment (jodh-intel) [reply src/agent/src/main.rs:206]: Thanks - this should be benign, but is unnecessary so I've removed 'em ;)

File: src/agent/src/tracer.rs (hunk 2 above)
Comment (jodh-intel) [reply src/agent/src/tracer.rs:74]: afaik no, but I've added a comment so it looks less odd.

File: src/trace-forwarder/src/handler.rs (hunk 3 above)
Comment (jodh-intel) [reply src/trace-forwarder/src/handler.rs]: Thanks @liubin - updated.

**Comment by jodh-intel** [comment]
Ping for reviews folk. Would be great to get this landed :smile:

**Review by devimc — APPROVED** [review]
thanks @jodh-intel

**Review by liubin — APPROVED** [review]

**Pull request merged**